Introduce mcdc::TVIdxBuilder (LLVM side, NFC)

chapuni opened the pull request:
This is a preparation of incoming Clang changes (#82448) and just checks `TVIdx` is calculated correctly. NFC.

`TVIdxBuilder` calculates deterministic Indices for each Condition Node. It is used for `clang` to emit `TestVector` indices (aka ID) and for `llvm-cov` to reconstruct `TestVectors`.

This includes the unittest `CoverageMappingTest.TVIdxBuilder`.

See also
https://discourse.llvm.org/t/rfc-coverage-new-algorithm-and-file-format-for-mc-dc/76798

chapuni committed:
Implement MCDCTVIdxBuilder (LLVM side)

This accepts current version of profdata. The output might be different.

See also
https://discourse.llvm.org/t/rfc-coverage-new-algorithm-and-file-format-for-mc-dc/76798

chapuni committed:
Implement MCDCTVIdxBuilder and MCDCTestVectorBuilder (LLVM side)

This accept current version of profdata. The output might be different.

See also
https://discourse.llvm.org/t/rfc-coverage-new-algorithm-and-file-format-for-mc-dc/76798

chapuni committed:
[Coverage] MCDCRecordProcessor: Find `ExecVectors` directly

Deprecate `TestVectors`, since no one uses it.

This affects the output order of ExecVectors.
The current impl emits sorted by binary value of ExecVector.
This impl emits along the traversal of `buildTestVector()`.

chapuni committed:
Revert "Implement MCDCTVIdxBuilder and MCDCTestVectorBuilder (LLVM side)"

This reverts commit d168e0cb85eb150caa7ab241f136c5a23f79ba38.

chapuni committed:
Merge branch 'mcdc/xv' into HEAD

chapuni commented:
I reworked to align the current implementation.

chapuni committed:
Update comments and assertions

chapuni committed:
Merge remote-tracking branch 'origin/main' into mcdc/tvidx

evodius96 commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -223,9 +223,127 @@ Expected<int64_t> CounterMappingContext::evaluate(const Counter &C) const {
+  Nodes[0].Width = 1;
+  Q.push_back(0);
+
+  unsigned Ord = 0;
```
Can you add a bit more commentary on the overall process of the algorithm (and MCDCTVIdxBuilder)?  What are the expected inputs and the expected output, generally?  I think it would help with the readability a bit more.

chapuni replied:
I've added comments and the unittest. Hope it helps!

evodius96 commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -283,35 +408,38 @@ class MCDCRecordProcessor {
+    const auto &Node = Nodes[ID];
+
+    for (unsigned I = 0; I < 2; ++I) {
+      auto MCDCCond = (I ? MCDCRecord::MCDC_True : MCDCRecord::MCDC_False);
```
Combining the True and False processing using a 2-iteration loop works (I had considered doing this too, since it's really doing the same thing for True and False), but it seems slightly less readable and a bit more complicated than what was there before.

evodius96 replied:
It's just a consideration, though. I'm OK with it.

chapuni replied:
See #81221

evodius96 commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -283,35 +408,38 @@ class MCDCRecordProcessor {
     MCDCRecord::TestVector TV(NumConditions, MCDCRecord::MCDC_DontCare);
-    buildTestVector(TV, 1, 0);
+    buildTestVector(TV, 0, 0, 0);
+    assert(TVIdxs.size() == NumTestVectors && "TVIdxs wasn't fulfilled");
```
Good assert

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -223,9 +223,127 @@ Expected<int64_t> CounterMappingContext::evaluate(const Counter &C) const {
+  // Build Nodes and set up each InCount
+  int MaxID = -1;
+  Nodes.resize(std::get<0>(Fetcher(true)));
+  while (true) {
```
I think it's fine for structured bindings to live inside loop headers.

```
int ID = 0;
for (auto [ID1, FalseID1, TrueID1] = Fetcher(false); ID1; ID = ID1 - 1;)
```

chapuni replied:
`Fetcher` is expected inside the loop.

Re: correlation between `ID1` and `ID`, see also #81257

chapuni replied:
I don't use `Fetcher` since I have introduced `NextIDsBuilder`.

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -223,9 +223,127 @@ Expected<int64_t> CounterMappingContext::evaluate(const Counter &C) const {
+    return;
+
+  // Sort key ordered by <-Width, Ord>
+  SmallVector<std::tuple<int,      /// -Width
```
why a `std::tuple` instead of struct?

chapuni replied:
I am using it as the sort key of comparator.

chapuni reviewed (commented):
Thanks for taking a look. I will update this evening.

ornata commented on llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h:
```diff
@@ -557,6 +558,35 @@ struct MCDCRecord {
+public:
+  struct MCDCNode {
+    int InCount = 0; /// Reference count; temporary use
+    int Width;       /// Number of paths (>= 1)
```
Why not name it `NumPaths`?

chapuni replied:
It was `W` in my prototype design. I have to update my imagination if this would be renamed.
Or, does the comment mislead?

ornata replied:
https://en.wikipedia.org/wiki/Width_(disambiguation) Width has many meanings for graphs. :)

chapuni replied:
Which is better for me to rename to easy `N` or to define `Width` here.
I haven't learnt the graph theory though, looks like I have introduced similar concept, as far as I read introductions of the theory.

Let me think more.

chapuni committed:
Reorganize TVIdxBuilder

* Split out `Indices[ID][Cond]`
* Let `Nodes` debug-only.
* Introduce `Offset`
* Introduce `HardMaxTVs`

chapuni committed:
Merge remote-tracking branch 'origin/main' into mcdc/tvidx

chapuni committed:
[CoverageMapping] Refactor `mcdc::TVIdxBuilder`

* Sink `TVIdxBuilder` into `mcdc::`.
* The ctor accepts `SmallVector<ConditionIDs>` indexed by `ID`.
* `class NextIDsBuilder` provides `NextIDs` as`SmallVector<ConditionIDs>`,
  for `TVIdxBuilder` to use it before `MCDCRecordProcessor()`.
  It was `BranchParamsMap` or `Map` as `DenseMap<Branch>`.
* `NodeIDs` and `Fetcher` function are deprecated.

chapuni committed:
Merge remote-tracking branch 'origin/main' into HEAD

chapuni committed:
remove <functional>

chapuni committed:
Update comments.

chapuni committed:
Add unittest

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -223,9 +223,130 @@ Expected<int64_t> CounterMappingContext::evaluate(const Counter &C) const {
+      Indices[ID][I] = NextNode.Width;
+      auto NextWidth = int64_t(NextNode.Width) + Node.Width;
+      if (NextWidth > HardMaxTVs) {
+        NumTestVectors = HardMaxTVs; // Overflow
```
Would it be possible to add debug output when there is an overflow?

chapuni replied:
I guess overflow is impossible in llvm-cov.
OTOH, Clang side will catch such a overflow and tell it with `HardMaxTVs` for warnings/errors. See `MCDCCoverageBuilder` in `CoverageMappingGen.cpp`, and `mcdc-error-conditions.cpp`.

I wonder it would be valuable to add debug message here.

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -223,9 +223,130 @@ Expected<int64_t> CounterMappingContext::evaluate(const Counter &C) const {
+    }
+  }
+
+  std::sort(Decisions.begin(), Decisions.end());
```
Can you use llvm::sort?

I think it takes ranges so you can just do `llvm::sort(Decisions)`

chapuni replied:
I've forgot it.

chapuni committed:
Use llvm::sort

chapuni committed:
EXPECT_

chapuni committed:
Merge remote-tracking branch 'chapuni/main' into mcdc/tvidx

ZhuUx commented on llvm/lib/ProfileData/Coverage/CoverageMapping.cpp:
```diff
@@ -223,9 +223,130 @@ Expected<int64_t> CounterMappingContext::evaluate(const Counter &C) const {
+    }
+  }
+
+  // Sort key ordered by <-Width, Ord>
```
Why end nodes of the decision need to be sorted in descending order of width?

chapuni replied:
https://discourse.llvm.org/t/rfc-coverage-new-algorithm-and-file-format-for-mc-dc/76798#other-minor-implementations-5

I think it may be available for integrating branch and mcdc.

See also;
https://discourse.llvm.org/t/rfc-region-branch-coverage-by-bitmap/79629#branch-coverage-and-mcdc-test-vectors-4

ZhuUx replied:
So motivation of the sort is to prepare for truncating all conditions with `width=1`, is it?
If so, why not just remove them then with an O(n) method?

chapuni replied:
Why did you think they could be removed? It's the prerequisite to assign identical IDs between clang and llvm-cov.

ZhuUx replied:
I see, if we did not sort them, we can not reduce size of bitmaps effectively since there might be other conditions with `width>1` getting large index.

I'm trying to remove the sort because I find it breaks some nature if the topological relationship of conditions is abnormal (e.g conditions from pattern matching in rust) and generates troubled index maps. I would investigate it more. Thanks!

chapuni replied:
Interesting. Could you introduce me the discussion, or paste the problematic cond tree anywhere?

ZhuUx replied:
:relaxed: Sorry I found I made a mistake. I previously encountered a example with decision tree:
```mermaid
flowchart TD
C0 -- T --> C1
C1 -- F --> F1
C1 -- T --> C3
C0 -- F --> C2
C2 -- T --> C3
C2 -- F --> F2
C3 -- F --> C4
C3 -- T --> T3
C4 -- T --> T4
C4 -- F --> F4
```
Normal boolean expressions can not generate such decision tree, where either true next of a condition is a next (true or false) of its false next, or false next of a condition is a next of its true next. 

I tried to disable the sort and fixed some cases. So I guessed something was broken by the sort. But that's wrong. The real problem is the way to update cond bitmap, which is specified for previous cond bitmap but is inappropriate for current. This implementation still works for pattern matching, thanks to your excellent efforts!

chapuni replied:
At a glance, `F1` and `F2` have `w=1`, and others have `w=2`. Correct?

ZhuUx replied:
Yes. The index map is right. The key difference is at most one of `C0` and `C2` can be true, which not likes `(C0 && C1) || C2`. In previous I updated false bit of `C2` when `C0` is true since if `C0` is true, `C2` must be false. While at present we can not do this.
